### ISSUE TITLE
question: home in sysroot deprecated?

### ISSUE THREAD

**Issue by agners** [ostreedev/ostree#2085]
Pull request #1894 added a `--modern` flag. It does get rid of all directories in `sysroot` except `boot` and `ostree`. Furthermore, the test `admin-test.sh` has been extended to check that `home` is not present (see https://github.com/ostreedev/ostree/pull/1894/files#diff-5092e9dd89d398d2a7b46be8e0dbe787). Is using home in sysroot deprecated/not recommended?

It is still the default option  documented in the docs currently https://ostree.readthedocs.io/en/latest/manual/adapting-existing/. Just wanted to get some clarification on the topic.

In case both are supported, is there a guidance in which case one would choose one over the other?

**Comment by cgwalters** [comment]
> In case both are supported, is there a guidance in which case one would choose one over the other?

I think both will be supported forever - OSTree tries hard to leave `/var` as "administrator territory" and never touch it.

The idea behind having `/home` in the physical root is that it's independent of any deployment; for example, you can use ostree to install *inside* an existing operating system and use `/home` in both places.

But, if you're not using OSTree in any kind of dual/multi-operating-system-boot scenario, it's probably cleanest to just keep it as `/var/home`.

Does that help?

**Comment by agners** [comment]
Yes, that is what I was looking for, thanks!

**Comment by agners** [comment]
@cgwalters I do have a follow up question still: Is it also a valid/supported option to commit the home directory? It seems to work, e.g. when I commit a home directory, it appears in the deployment as regular directory and files in it are hardlinked to object in the repo. I guess it might be an use-case in a "userless" embedded system.

**Comment by cgwalters** [comment]
>  I do have a follow up question still: Is it also a valid/supported option to commit the home directory? It seems to work, e.g. when I commit a home directory, it appears in the deployment as regular directory and files in it are hardlinked to object in the repo. I guess it might be an use-case in a "userless" embedded system.

No, please don't do it.  If you want a home directory to come with content from the OS, a good approach is to use the default `/etc/skel` model - and notably invoke `useradd` on the client, not the server.

When doing things like this you need to think about how upgrades work.  While today OSTree doesn't strictly enforce that you don't have content in `/var`...I would like to make it an error in the future.  

I really don't want `ostree admin rollback` to break user data *ever* - and having OSTree write data into the home directory can lead to that.

**Comment by agners** [comment]
> When doing things like this you need to think about how upgrades work.

Well I guess in my case I would make `/home` read-only just like `/usr`. I mean it would be a really special case for embedded system without any real user in that sense.

> While today OSTree doesn't strictly enforce that you don't have content in /var...I would like to make it an error in the future.

Today in `meta-updater` `/sysroot/home` is prepopulated with the users already configured at OE build time. With the move to `/var/home` this of course means pre-populated `/var`... This of course will make creating a new deployment more cumbersome than just `ostree admin os-init`.

One way I was thinking first was to have `/usr/home` committed as template (similar to `/usr/etc`). But not shipping with any user but with (first-boot) scripts which setup user seems like a good way here I agree. I will investigate how to do this in OpenEmbedded. Are you aware of such infrastructure in systemd or the likes?

> I really don't want ostree admin rollback to break user data ever - and having OSTree write data into the home directory can lead to that.

```
error: Unknown "admin" subcommand 'rollback'
```

**Comment by cgwalters** [comment]
> Well I guess in my case I would make /home read-only just like /usr

Hmm.  How about instead shipping the users in `/etc/passwd` with their home directory set to `/usr/home/user1` e.g.?   Then it's all clear how it works - we already have the read-only bind mount there.  The files can be owned by root because they're immutable, etc.

> error: Unknown "admin" subcommand 'rollback'

Right sorry, need to migrate that logic from `rpm-ostree`.

**Comment by agners** [comment]
> Hmm. How about instead shipping the users in /etc/passwd with their home directory set to /usr/home/user1 e.g.? Then it's all clear how it works - we already have the read-only bind mount there. The files can be owned by root because they're immutable, etc.

That indeed sounds like the cleaner approach.

**Comment by cgwalters** [comment]
BTW, you could also consider using [systemd dynamic users](http://0pointer.net/blog/dynamic-users-with-systemd.html).
